File: openemm-backend/src/main/c/lib/node.c
/*	-*- mode: c; mode: fold -*-	*/
/*********************************************************************************
 * The contents of this file are subject to the Common Public Attribution
 * License Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.openemm.org/cpal1.html. The License is based on the Mozilla
 * Public License Version 1.1 but Sections 14 and 15 have been added to cover
 * use of software over a computer network and provide for limited attribution
 * for the Original Developer. In addition, Exhibit A has been modified to be
 * consistent with Exhibit B.
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for
 * the specific language governing rights and limitations under the License.
 * 
 * The Original Code is OpenEMM.
 * The Original Developer is the Initial Developer.
 * The Initial Developer of the Original Code is AGNITAS AG. All portions of
 * the code written by AGNITAS AG are Copyright (c) 2007 AGNITAS AG. All Rights
 * Reserved.
 * 
 * Contributor(s): AGNITAS AG. 
 ********************************************************************************/
/** @file node.c
 * Handle nodes in hash collection.
 */
# include	<stdlib.h>
# include	<string.h>
# include	"agn.h"

/** Allocate a generic node.
 * @param key the key
 * @param klen the length of the key
 * @param hash the hash code
 * @param data the data
 * @param dlen the length of the data
 * @return the new node on success, NULL otherwise
 */
gnode_t *
gnode_alloc (const byte_t *key, int klen, hash_t hash,
	     const byte_t *data, int dlen) /*{{{*/
{
	gnode_t	*g;
	
	if (g = (gnode_t *) malloc (sizeof (gnode_t))) {
		g -> key = (byte_t *) malloc (klen + 1);
		g -> klen = klen;
		g -> hash = hash;
		g -> data = (byte_t *) malloc (dlen + 1);
		g -> dlen = dlen;
		g -> next = NULL;
		if (g -> key && g -> data) {
			if (klen > 0)
				memcpy (g -> key, key, klen);
			if (dlen > 0)
				memcpy (g -> data, data, dlen);
		} else
			g = gnode_free (g);
	}
	return g;
}/*}}}*/
/** Frees generic node.
 * @param g the node to free
 * @return NULL
 */
gnode_t *
gnode_free (gnode_t *g) /*{{{*/
{
	if (g) {
		if (g -> key)
			free (g -> key);
		if (g -> data)
			free (g -> data);
		free (g);
	}
	return NULL;
}/*}}}*/
/** Frees generic nodes.
 * @param g the node to start
 * @return NULL
 */
gnode_t *
gnode_free_all (gnode_t *g) /*{{{*/
{
	gnode_t	*tmp;
	
	while (tmp = g) {
		g = g -> next;
		gnode_free (tmp);
	}
	return NULL;
}/*}}}*/
/** Set a generics node data.
 * @param g the node to update
 * @param data the data to use
 * @param dlen the length of the data
 */
bool_t
gnode_setdata (gnode_t *g, const byte_t *data, int dlen) /*{{{*/
{
	bool_t	rc;
	byte_t	*tmp;
	
	if ((g -> dlen >= dlen) && g -> data) {
		rc = true;
	} else if (tmp = realloc (g -> data, dlen + 1)) {
		g -> data = tmp;
		rc = true;
	} else
		rc = false;
	if (rc) {
		memcpy (g -> data, data, dlen);
		g -> dlen = dlen;
	}
	return rc;
}/*}}}*/

/** Allocate node.
 * Allocate memory for a new node
 * @param mkey the key for the hashing
 * @param hash the hash code
 * @param okey the original key
 * @param data the value for this node
 * @return the new node on success, NULL otherwise
 */
node_t *
node_alloc (const char *mkey, hash_t hash,
	    const char *okey, const char *data) /*{{{*/
{
	node_t	*n;
	
	if (n = (node_t *) malloc (sizeof (node_t))) {
		n -> mkey = NULL;
		n -> hash = hash;
		n -> okey = NULL;
		n -> data = NULL;
		n -> next = NULL;
		if ((mkey && (! (n -> mkey = strdup (mkey)))) ||
		    (okey && (! (n -> okey = strdup (okey)))) ||
		    (data && (! (n -> data = strdup (data)))))
			n = node_free (n);
	}
	return n;
}/*}}}*/
/** Frees node.
 * Return the memory allocated to the system
 * @param n the node to free
 * @return NULL
 */
node_t *
node_free (node_t *n) /*{{{*/
{
	if (n) {
		if (n -> mkey)
			free (n -> mkey);
		if (n -> okey)
			free (n -> okey);
		if (n -> data)
			free (n -> data);
		free (n);
	}
	return NULL;
}/*}}}*/
/** Free nodes.
 * Return the resources of the node and all siblings to the system
 * @param n the node to start
 * @return NULL
 */
node_t *
node_free_all (node_t *n) /*{{{*/
{
	node_t	*tmp;
	
	while (tmp = n) {
		n = n -> next;
		node_free (tmp);
	}
	return NULL;
}/*}}}*/
/** Set node data.
 * Set/change the value for this node
 * @param n the node to change
 * @param data the new content
 * @return true on success, false otherwise
 */
bool_t
node_setdata (node_t *n, const char *data) /*{{{*/
{
	if (n -> data)
		free (n -> data);
	n -> data = data ? strdup (data) : NULL;
	return (data && (! n -> data)) ? false : true;
}/*}}}*/

File: openemm/src/main/java/org/agnitas/service/csv/Node.java
/*********************************************************************************
 * The contents of this file are subject to the Common Public Attribution
 * License Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.openemm.org/cpal1.html. The License is based on the Mozilla
 * Public License Version 1.1 but Sections 14 and 15 have been added to cover
 * use of software over a computer network and provide for limited attribution
 * for the Original Developer. In addition, Exhibit A has been modified to be
 * consistent with Exhibit B.
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for
 * the specific language governing rights and limitations under the License.
 *
 * The Original Code is OpenEMM.
 * The Original Developer is the Initial Developer.
 * The Initial Developer of the Original Code is AGNITAS AG. All portions of
 * the code written by AGNITAS AG are Copyright (c) 2009 AGNITAS AG. All Rights
 * Reserved.
 *
 * Contributor(s): AGNITAS AG.
 ********************************************************************************/

package org.agnitas.service.csv;

/**
 * Node used by <code>BeanBuilder</code> to track the pushing/popping of bean properties.
 * <p>
 * A node is primarily defined by:
 * <ul>
 * <li>context - the target of the node, ie the JavaBean/collection onto which node operations will be applied</li>
 * <li>name - the name of the node, ie the JavaBean property name dictating how node operations will be applied to the
 * <i>parent</i> context</li>
 * </p>
 * <p>
 * Nodes have basic operations performed on them:
 * <ul>
 * <li>A text value can be set - this will typically affect the node's value although for nodes representing beans or
 * collections this will not be the case</li>
 * <li>A value can be retrieved - for JavaBean/collection nodes this will be the target bean or collection, for other nodes
 * this will probably be dependent on the textual value previously set</li>
 * <li>A value can be applied to the context - for JavaBean nodes this will set a property of the JavaBean, for collections
 * this will add a value to that collection</li>
 * </ul>
 * </p>
 * <p>
 * The methods of a node correspond to BeanBuilder events as follows:
 * <ul>
 * <li>push - an appropriate node type is created</li>
 * <li>apply - the text value is set on the node (and an appropriate value created if necessary)</li>
 * <li>pop - the node is popped from the head of the stack and its value is applied to <i>its parent node</i></li>
 * </ul>
 * </p>
 *
 * @author Viktor Gema
 */
public abstract class Node {

    private final Object context;

    private final String name;

    protected Node(Object context, String name) {
        this.context = context;
        this.name = name;
    }

    public Object getContext() {
        return context;
    }

    public String getName() {
        return name;
    }

    /**
     * Sets the value of this node via a text value.
     *
     * @param text The text value used to set this node's value
     * @throws Exception Thrown if the text could not be applied to this node, eg if the text is not in a suitable format
     */
    public abstract void setText(String text) throws Exception;

    /**
     * The value of this node.
     * <p>
     * This may have been set by a call to {@link #setText}.
     * </p>
     *
     * @return Returns the value of this node, may be <code>null</code>
     */
    public abstract Object getValue();

    /**
     * Applies a named value to this node.
     * <p>
     * This represents a child node being popped after it has finished being constructed. For example, a JavaBean might apply a
     * property value like this whereas a Collection might simply add the value, regardless of the name.
     * </p>
     *
     * @param name  The name of the child node that was popped
     * @param value The value of the child node
     * @throws Exception Thrown if the named value can't be applied to this node
     */
    public abstract void apply(String name, Object value) throws Exception;
}
